src/stdpar/plugin-SiPixelClusterizer/gpuClustering.h

```diff
@@ -1,292 +1,262 @@
 
+#include <algorithm>
 #include <cstdint>
 #include <cstdio>
+#include <execution>
+#include <iterator>
+#include <ranges>
 
 #include "Geometry/phase1PixelTopology.h"
 #include "CUDACore/HistoContainer.h"
-#include "CUDACore/cuda_assert.h"
+#include "CUDACore/portableAtomicOp.h"
 
 #include "gpuClusteringConstants.h"
 
 namespace gpuClustering {
 
 #ifdef GPU_DEBUG
-  __device__ uint32_t gMaxHit = 0;
+  uint32_t gMaxHit = 0;
 #endif
 
-  __global__ void countModules(uint16_t const* __restrict__ id,
+  void countModules(uint16_t const* __restrict__ id,
                                uint32_t* __restrict__ moduleStart,
                                int32_t* __restrict__ clusterId,
                                int numElements) {
-    int first = blockDim.x * blockIdx.x + threadIdx.x;
-    for (int i = first; i < numElements; i += gridDim.x * blockDim.x) {
+    auto iter{std::views::iota(0, numElements)};
+    std::for_each(std::execution::par, std::ranges::cbegin(iter), std::ranges::cend(iter), [=](const auto i) {
       clusterId[i] = i;
-      if (InvId == id[i])
-        continue;
-      auto j = i - 1;
-      while (j >= 0 and id[j] == InvId)
-        --j;
-      if (j < 0 or id[j] != id[i]) {
-        // boundary...
-        auto loc = atomicInc(moduleStart, MaxNumModules);
-        moduleStart[loc + 1] = i;
+      if (InvId != id[i]){
+        auto j = i - 1;
+        while (j >= 0 and id[j] == InvId)
+          --j;
+        if (j < 0 or id[j] != id[i]) {
+          // boundary...
+          auto loc = cms::cuda::atomicInc(moduleStart, MaxNumModules);
+          moduleStart[loc + 1] = i;
+        }
       }
-    }
+    });
   }
 
-  __global__
-      //  __launch_bounds__(256,4)
-      void
-      findClus(uint16_t const* __restrict__ id,           // module id of each pixel
+  void findClus(uint16_t const* __restrict__ id,           // module id of each pixel
                uint16_t const* __restrict__ x,            // local coordinates of each pixel
                uint16_t const* __restrict__ y,            //
                uint32_t const* __restrict__ moduleStart,  // index of the first pixel of each module
                uint32_t* __restrict__ nClustersInModule,  // output: number of clusters found in each module
                uint32_t* __restrict__ moduleId,           // output: module id of each module
                int32_t* __restrict__ clusterId,           // output: cluster id of each pixel
                int numElements) {
-    if (blockIdx.x >= moduleStart[0])
-      return;
 
-    auto firstPixel = moduleStart[1 + blockIdx.x];
-    auto thisModuleId = id[firstPixel];
-    assert(thisModuleId < MaxNumModules);
+    uint32_t firstModule = 0;
+    uint32_t endModule = moduleStart[0];
+    auto iter{std::views::iota(firstModule, endModule)};
+    std::for_each(std::execution::par, std::ranges::cbegin(iter), std::ranges::cend(iter), [=](const auto module) {
```

Comment:
This Kernel has an outer loop iterating over modules and multiple inner loop within a single module iteration. Since we can't have nested parallel algorithms I think parallelizing the outer loop should be more efficient, inner loop are smaller and would result in much more kernel launch